Update handling forms - Controlled Forms (#135)
Ada-Developers-Academy/core-unit-3: handling-forms/controlled-forms.md

Reviewer: In the `06-handling-forms-npm-id` branch of `student-app` I made some changes to the form because the browser was complaining that we were using `htmlFor` but the inputs didn't have `id`s; they only had `name`s.

I added `id`s, so the `NewStudentForm` snippets showing JSX should be updated to use them (I count 4 copies of the NewStudentForm JSX), and the callout that talks about htmlFor should probably mention `id` and how `for` requires `an `id` for proper functionality.

It's probably to also point out that using `id` in components can be dangerous from the perspective that if there were two instances of the component placed on the page, we would have two elements with the same id (a no-no), but for the form controls we usually only expect a single control. There could be a FYC blurb for folks who are wondering how we could deal with cases where we need an id AND multiple component copies.

Reviewer: There's one mention of `NewStudentForm.js` that should be `NewStudentForm.jsx`

Reviewer: What would you think about adding a callout for a Refactoring Opportunity to use a combined change handler rather than individual change handlers. We end up showing it in roundtables, but I don't see it in Learn.

There's the slightly confusing point that the name used on each field needs to match the state key (email does, fullName does not, for some reason). But it would give us a place to actually show the computed property names syntax.

Author: Yeah I can work that into this PR

```js
import { useState } from 'react';
import StudentList from './components/StudentList';
import ClassInfo from './components/ClassInfo';

function App() {
  const [studentData, setStudentData] = useState([
    {
      id: 1,
      nameData: 'Ada',
      emailData: 'ada@dev.org',
      isPresentData: false,
    },
    {
      id: 2,
      nameData: 'Soo-ah',
      emailData: 'sooah@dev.org',
      isPresentData: false,
    },
    {
      id: 3,
      nameData: 'Chrissy',
      emailData: 'chrissy@dev.org',
      isPresentData: true,
    },
  ]);

  const toggleStudentPresence = (studentId) => {

>     // calculate the updated student data by finding the student that matches

Reviewer: I've been removing all the comments when re-presenting the state-of-the-app code.

    // the passed id, making a copy with object spreading, then overwriting
    // the presence value with its inverse
    const students = studentData.map(student => {
      if (student.id === studentId) {
        // this was the toggled student, so make a new record with the updated
        // presence value
        return { ...student, isPresentData: !student.isPresentData };
      } else {
        // this was not the student who was toggled, so we can use the existing
        // data in the new student array
        return student;
      }
    });

    // uses value-passing style to update the student data, but could be
    // refactored to use function-passing style
    setStudentData(students);
  };

  return (
    <main>
      <h1>Attendance</h1>

>       <ClassInfo memberCount={studentData.length}></ClassInfo>

Reviewer: remove ClassInfo code from the examples
```suggestion
```

      <StudentList
        students={studentData}
        onStudentPresenceToggle={toggleStudentPresence}
      ></StudentList>
    </main>

>   );

Author: Pulled this code from: https://github.com/AdaGold/student-app/blob/05-lifting-state-up-npm-id/src/App.jsx

Reviewer: Though the repo code is using 2 spaces, Learn was using 4, so when I copied the new code into other lessons I ran a quick search/replace over the code lines to replace all occurrences of 2 spaces with 4 spaces

<!-- prettier-ignore-start -->
```js

> import Student from './Student';

Reviewer: In other lessons, I left this line where it was before to minimize the diff

import './StudentList.css';
import PropTypes from 'prop-types';

const StudentList = (props) => {

>   const studentComponents = props.students.map(student => {

Reviewer: 2 space → 4 space replacement throughout the rest of this "file".

<!-- prettier-ignore-start -->
```js
import './Student.css';

> import PropTypes from 'prop-types';

Reviewer: Here too, I would tend to leave the line where it was to minmize the diff.

const Student = (props) => {

>   const attendanceButtonClicked = () => {

Reviewer: 2 space → 4 space replacement throughout the rest of this "file".

>     // Invoke the function passed in through the prop named "onPresenceToggle"
>     // This function refers to the toggleStudentPresence function in App

Reviewer: I'd remove these comments
```suggestion
```

<br/>

The `id` and `for` attributes enable us to [associate a `<label>` element with its corresponding `<input>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#labels_2) which improves accessibility because assistive technologies can tell what the input is for and the area a user can click is increased from just the input element to include the label element too. 

<br/>

The [name attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#name) on an `<input>` element assigns a name to tit and this name can be used during form submission or as a reference for other purposes.

<br/>

We know that we need to include the `id` attribute for accessibility reasons, but what if we want to use the same component on a page two times? If the same component appears in an application more than once then the `id`, which is meant to be a unique identifier, would also show up more than once. In this scenario, the `id` would not be unique in the entire DOM.

<br/>

Expand below to learn how we can ensure `id`s are unique in a React app.

<br/>

> <details>

Author: Wasn't sure if this infor should be in a <detail> or not but the callout felt like it was getting long. What do you think?

## Check for Understanding

<!-- Question 1 -->
<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: 2c36b6ce-cc23-455c-b6f5-41de229d3a05 
* title: Controlled Forms
##### !question

Controlled components follow a design pattern that states that...

##### !end-question
##### !options

a| A form's values should be controlled by the component's state, not by user interaction.
b| A form's values should be controlled by user interaction, not by the component's state.
c| A form's values should be controlled by what a user types into a form's input element.

##### !end-options
##### !answer

a|

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 2 -->
<!-- prettier-ignore-start -->
### !challenge
* type: ordering
* id: 114a72bf-6850-4e12-aae7-3195d372fc5d
* title: Controlled Forms
##### !question

> You are helping a dog rescue organization add a form to their website to collect applicants' information. Arrange the following steps for creating a controlled component called `AdoptionApplicationForm`.

Reviewer: Animal rescue? Don't leave out the kitties!
```suggestion
You are helping an animal rescue organization add a form to their website to collect applicants' information. Arrange the following steps for creating a controlled component called `AdoptionApplicationForm`.
```

Since we will use event data in our `handleChange`, we will also need to make a slight update to the attributes for the `label` and `input` elements. When we use the event data to dynamically get the user input, we use `event.target.name` to populate the property for the `formFields` object and we use `event.target.value` to populate what the user typed as the value.

<br/>

If we the attributes for the `label` and `input` elements as `fullName` then `event.target.name` would evaluate to the string `fullName` because the `name` attribute was set to the value "fullName".

<br/>

This is where we run into an issue. The formFields state object has a key "name" not "fullName" so when we use `formFields.name` to have our form input read from the updated state, we would get `undefined` because `handleChange` would have set the property to "fullName". Thus, we need to update `input`'s `name` attribute to be the string "name" instead of "fullName". 

<br/>

> Additionally, to associate the `input` with the `label` we then need to also change `id` and `for` attributes to be "name" instead of "fullName".

Author: This feels pretty lengthy and might not click with students until they see it for themselves. Any feedback here would be helpful. Thanks!!

> Since we will use event data in our `handleChange`, we will also need to make a slight update to the attributes for the `label` and `input` elements. When we use the event data to dynamically get the user input, we use `event.target.name` to populate the property for the `formFields` object and we use `event.target.value` to populate what the user typed as the value.
> 
> <br/>
> 
> If we the attributes for the `label` and `input` elements as `fullName` then `event.target.name` would evaluate to the string `fullName` because the `name` attribute was set to the value "fullName".
> 
> <br/>
> 
> This is where we run into an issue. The formFields state object has a key "name" not "fullName" so when we use `formFields.name` to have our form input read from the updated state, we would get `undefined` because `handleChange` would have set the property to "fullName". Thus, we need to update `input`'s `name` attribute to be the string "name" instead of "fullName". 
> 
> <br/>
> 
> Additionally, to associate the `input` with the `label` we then need to also change `id` and `for` attributes to be "name" instead of "fullName".

Reviewer: Somewhat aggressive condensing of the discussion.
```suggestion
In the unified `handleChange` function below, we access the `name` attribute of the changed `input` as `event.target.name`. This will be used as the key we update in our form state. We used the key `"name"` in our state, but notice that we originally used `"fullName"` as the `name` of our related `input`. This was primarily done to reduce the possibility of confusion between all the things called `name` and the value `"fullName"` in explanations.

<br />

When we use a unified change handler and get the key name from the changed `input` `name`, then we should plan to use the same state key name as the `input` `name` attribute. We would probably then also use similar names for the `id` and `forHtml` attributes for consistency as shown.
```

### !end-callout

### !callout-info

## Refactoring Opportunity: Combining Multiple Change Handlers into One

Notice that the functions `onNameChange` and `onEmailChange` are nearly identical. The only difference between the two event handlers is that one updates the value for the name property and one updates the value for the email property in the formFields state object.

<br/>

> We can refactor our code to combine  `onNameChange` and `onEmailChange` into one event handler called `handleChange` to reduce repetition in the `NewStudentForm` component.

Reviewer: Per my comment below.

```suggestion
We can refactor our code to combine `onNameChange` and `onEmailChange` into one event handler called `handleChange` to reduce repetition in the `NewStudentForm` component. `handleChange` will use details from the change event to figure out which part of the state to update!

<br />

Expand the section below for an explanation of a possible approach!
```

<br/>

> We can accomplish this by using the spread operator notation with object shorthand notation as we saw in the lesson on "Passing Down Event Handlers".

Author: This callout directly follows the callout above and visually takes up a lot of space. Maybe I should put the code in a detail to make this shorter

Reviewer: Yeah, for the other ones I had done, the change wasn't relying on such new concepts, so I could give the reader a "think about how to do x" and then provide the solution collapsed. I think here, you're going to need to hint at the overall apporach in the main callout, then have the detailed explanation and code in the detail. Your currently summary (through "We can refactor our code to combine...") sounds good, but I'd like it to mention that the state key to be updated will come from the event data. Added a suggestion above.

> We can accomplish this by using the spread operator notation with object shorthand notation as we saw in the lesson on "Passing Down Event Handlers".

Reviewer: I think the use of computed property notation in conjunction with the spread is what makes this tricky and should be mentioned.

```suggestion
We can accomplish this by using the spread operator notation with object shorthand notation as we saw in the lesson on "Passing Down Event Handlers". In that lesson, the key being updated was known ahead of time, and could be hard-coded as part of the shorthand notation. Here, the key is read from the event data, and must be computed each time we get a new change event. To do this with shorthand notation, we use [`computed property notation`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names). Essentially, the expression to calculate the key is placed within brackets `[]`.
```

```js
import { useState } from 'react';

const NewStudentForm = () => {

    const [formFields, setFormFields] = useState({
        name: '',
        email: ''

>     });

Reviewer: `fullName`/`name` is a problem here. We need to mention that the `input` `name` needs to match the state key name so that we can use the `name` to figure out the key. Since the `name` key is what's used in a student record, we'd probably change the `name` of the input to match, which gives us the potentially confusing `name="name"`, which is I'm sure why the original example went with `fullName` instead. But we should at least mention all this.
